src/invite-api.js

```diff
@@ -1,231 +1,391 @@
+   * those invites are removed, and the invitors are told that you're already
+   * part of this project.
+   *
+   * @param {Pick<Invite, 'inviteId'>} invite
    * @returns {Promise<void>}
    */
-  async accept(projectPublicId) {
-    const projectId = this.#getProjectIdFromPublicId(projectPublicId)
+  async accept({ inviteId: inviteIdString }) {
+    const inviteId = Buffer.from(inviteIdString, 'hex')
 
-    const isAlreadyMember = this.#isMember(projectId)
-    const peersToRespondTo = this.#peersToRespondTo.get(projectId)
+    const pendingInvite = this.#pendingInvites.getByInviteId(inviteId)
+    if (!pendingInvite) {
+      throw new Error(`Cannot find invite ID ${inviteIdString}`)
+    }
+
+    const { peerId, invite } = pendingInvite
+    const { projectName, projectPublicId } = invite
+
+    const removePendingInvite = () => {
+      const didDelete = this.#pendingInvites.deleteByInviteId(inviteId)
+      if (didDelete) this.emit('invite-removed', internalToExternal(invite))
+    }
 
+    // This is probably impossible in the UI, but it's theoretically possible
+    // to join a project while an invite is pending, so we need to check this.
+    const isAlreadyMember = this.#isMember(projectPublicId)
     if (isAlreadyMember) {
-      for (const peerId of peersToRespondTo) {
-        await this.#sendAlreadyResponse({ peerId, projectId })
+      const pendingInvitesDeleted =
+        this.#pendingInvites.deleteByProjectPublicId(projectPublicId)
+      for (const pendingInvite of pendingInvitesDeleted) {
+        this.#sendAlreadyResponse({
+          peerId: pendingInvite.peerId,
+          inviteId: pendingInvite.invite.inviteId,
+        })
+        this.emit('invite-removed', internalToExternal(pendingInvite.invite))
       }
       return
     }
 
-    const pendingInvite = this.#pendingInvites.get(projectId)
-
-    if (!pendingInvite) {
-      throw new Error(
-        `Cannot find invite for project with ID ${projectPublicId}`
-      )
-    }
+    assert(
+      !this.#pendingInvites.isAcceptingForProject(projectPublicId),
+      `Cannot double-accept invite for project ${projectPublicId}`
+    )
+    this.#pendingInvites.markAccepting(inviteId)
+
+    const projectDetailsAbortController = new AbortController()
+
+    const projectDetailsPromise =
+      /** @type {typeof onceSatisfied<TypedEmitter<import('./local-peers.js').LocalPeersEvents>, 'got-project-details'>} */ (
+        onceSatisfied
+      )(
+        this.rpc,
+        'got-project-details',
+        (projectDetailsPeerId, details) =>
+          // This peer ID check is probably superfluous because the invite ID
+          // should be impossible to guess, but is easy to add for correctness.
+          timingSafeEqual(projectDetailsPeerId, peerId) &&
+          timingSafeEqual(inviteId, details.inviteId),
+        { signal: projectDetailsAbortController.signal }
+      ).then((args) => args?.[1])
 
```

Comment:
I think this is needed to avoid an uncaught error if `#sendAcceptResponse` below throws, since this could throw in a tick after that once this function context no longer exists.

```suggestion

projectDetailsPromise.catch(noop)
```

Reply:
(we do want `projectDetailsProject` to be able to throw, hence this is on a separate line)

Reply:
I can easily change this, but `onceSatisfied` should never reject, even if aborted—it will resolve `null`. Therefore, I don't _think_ this needs to be caught. Happy to change any of that, including my understanding.

Reply:
ah, right, I hadn't fully parsed the implementation of `onceSatisfied`. I see it won't reject, but notice it can result in an uncaught error, if the `check` function throws, e.g. in this case it would throw if somehow `projectDetailsPeerId` or other params were a `number` or `undefined` (`timingSafeEqual` would throw a `TypeError`).

I think it's always best to assume that async functions throw, and make sure that they are caught, rather than relying on the assumption that they won't throw, because down the line it's easy to add some code that results in them throwing, and it's hard to retain the memory of "this function must never throw".

In a case like this I'm not a huge fan of returning `null` to indicate abort or an error from a helper like this, although I think that's just a personal preference. I prefer to throw, and if I want to use `null` as my check then I can do `fnThatMightThrow().catch(() => null)` or I can choose to handle the error a different way.

Reply:
TL;DR, `onceSatisfied` needs to wrap `check()` in a try...catch, and either reject or return `null` if it throws. My preference is to throw, but fine with keeping `res(null)` if preferred. I think we should have the `.catch(noop)` anyway, so anyone refactoring / changing code doesn't have to remember which async functions can throw and which can't.

Reply:
`onceSatisfied` now throws. I added `.catch(noop)` to its calls. See 759e5c5cf90daf00d346c34df35ec3768420b4e4.